plugins/tactics/src/Ide/Plugin/Tactic/Naming.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE ViewPatterns #-}
+
+module Ide.Plugin.Tactic.Naming where
+
+import           Control.Monad.State.Strict
+import           Data.Bool (bool)
+import           Data.Char
+import           Data.Map (Map)
+import qualified Data.Map as M
+import           Data.Traversable
+import           Name
+import           TcType
+import           TyCon
+import           Type
+import           TysWiredIn (listTyCon, pairTyCon, unitTyCon)
+
+
+------------------------------------------------------------------------------
+-- | Use type information to create a reasonable name.
+mkTyName :: Type -> String
+-- eg. mkTyName (a -> B) = "fab"
+mkTyName (tcSplitFunTys -> ([a@(isFunTy -> False)], b))
+  = "f" ++ mkTyName a ++ mkTyName b
+-- eg. mkTyName (a -> b -> C) = "f_C"
+mkTyName (tcSplitFunTys -> ((_:_), b))
+  = "f_" ++ mkTyName b
+-- eg. mkTyName (Either A B) = "eab"
+mkTyName (splitTyConApp_maybe -> Just (c, args))
+  = mkTyConName c ++ foldMap mkTyName args
+-- eg. mkTyName a = "a"
+mkTyName (getTyVar_maybe -> Just tv)
+  = occNameString $ occName tv
+-- eg. mkTyName (forall x. y) = "y"
+mkTyName (tcSplitSigmaTy -> ((_:_), _, t))
+  = mkTyName t
+mkTyName _ = "x"
+
+
+------------------------------------------------------------------------------
+-- | Get a good name for a type constructor.
+mkTyConName :: TyCon -> String
+mkTyConName tc
+  | tc == listTyCon = "l_"
+  | tc == pairTyCon = "p_"
+  | tc == unitTyCon = "unit"
+  | otherwise
+      = take 1
+      . fmap toLower
+      . filterReplace isSymbol      's'
+      . filterReplace isPunctuation 'p'
```

Comment:
replacing symbols and punctuation like this doesn't really need like it will lead to good names.

Reply:
No, but without it we run into problems giving names for type operators; it's really hard to come up with a name for a tycon called `(:<!>)`, eg.